hatchery/src/snapshot.rs

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::error::Error;
+use crate::storage_helpers::{
+    combine_module_snapshot_names, snapshot_id_to_name,
+};
+use crate::Error::PersistenceError;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use rkyv::{Archive, Deserialize, Serialize};
+pub const SNAPSHOT_ID_BYTES: usize = 32;
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Archive,
+    Serialize,
+    Deserialize,
+    PartialOrd,
+    Ord,
+    Hash,
+    Clone,
+    Copy,
+)]
+#[repr(C)]
+pub struct SnapshotId([u8; SNAPSHOT_ID_BYTES]);
+impl SnapshotId {
+    pub const fn uninitialized() -> Self {
+        SnapshotId([0u8; SNAPSHOT_ID_BYTES])
+    }
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+}
+impl From<[u8; 32]> for SnapshotId {
+    fn from(array: [u8; 32]) -> Self {
+        SnapshotId(array)
+    }
+}
+
+pub trait SnapshotLike {
+    fn path(&self) -> &PathBuf;
+    /// Read's snapshot's content into buffer
+    fn read(&self) -> Result<Vec<u8>, Error> {
+        let mut f = std::fs::File::open(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let metadata = std::fs::metadata(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let mut buffer = vec![0; metadata.len() as usize];
+        f.read(buffer.as_mut_slice()).map_err(PersistenceError)?;
+        Ok(buffer)
+    }
+}
+
+pub struct MemoryPath {
```

Comment:
This structure can be removed if you implement `SnapshotLike` directly on `Path`

Reply:
But I wanted to have MemoryPath as a type, to pass it safely around and distinguish it from Path.
You suggest to do `type MemoryPath = Path` instead?

Reply:
If you want to keep a type alias, sure. It seems that you could just distinguish it by name though.

Reply:
Issue is that it is very easy to make mistake, MemoryPath is something that is a current live memory for a module, to be distinguished from just passive snapshot. To avoid confusion in parameters, I wanted to keep this type really separate.

Reply:
This is called 'newtype' pattern - wrapping a core type within a single field struct (or a tuple)